### ISSUE TITLE
Executing BASH command and printing that command substitution

### ISSUE THREAD

**Issue by fxquah** [fxquah/QandA#2]
Suppose I have the following bash script `main.sh`:

```
#!/bin/bash

A=fileA
COMMAND="$(touch $A)"
```

I want a line(s) of code that does both of these:
* executes `$COMMAND` (create fileA).
* echoes/prints the command which was run (`touch fileA`), after substituting the value of `$A`

### What I tried
`eval "${COMMAND}"` : does the first thing only (creates the file).

`echo "${COMMAND}"` : does the first thing, and outputs the stdout (which in this case, an empty line), but not the actual command which was run.

Any advice?

**Comment by fxquah** [comment]
Even if there isn't a way to do both things in one line, I still want to know how to do the second task (i.e. printing the command as it was run).

**Comment by shouldsee** [comment]
```bash
#!/bin/bash

A=fileA
COMMAND="touch $A"
eval "$COMMAND"
```

**Comment by shouldsee** [comment]
The first thing would already been done once you issue

```
#!/bin/bash

A=fileA
COMMAND="$(touch $A)"
```

Because `$(touch $A)` means "execute the command `touch $A`" and capture its output as a string(loosely)

**Comment by fxquah** [comment]
What I want to see is this line
```
touch fileA
```
which I will put in my log file of what commands were run.


```
echo "[CMD]$CMD"
[[ "$DRY" -eq 1 ]] || eval "$COMMAND"
```
This one does not work for me. Even after I change `CMD` to `COMMAND`. What it does is to echo:
```[CMD]main.sh```
the second part is the list of file in the directory.

**Comment by fxquah** [comment]
```
echo "[CMD]$CMD"
[[ "$DRY" -eq 1 ]] || eval "$COMMAND"
```
This one does not work for me. Even after I change `CMD` to `COMMAND`. What it does is to echo:
```[CMD]main.sh```
the second part is the list of file in the directory.

**Comment by shouldsee** [comment]
There was a typo in the previous version. Note the varaible storing command is $CMD in my snippet
```bash
#!/bin/bash

A=fileA
CMD="touch $A"

### This is the snippet that I reuse.
### It always echoes CMD, but skip its executation if $DRY is set to 1
echo "[CMD]$CMD"
[[ "$DRY" -eq 1 ]] || eval "$CMD"
```

**Comment by fxquah** [comment]
Yes it does, my mistake was not realising you removed the braces from the second line:
```
CMD="$(touch $A)"    # original
CMD="touch $A"       # changed
```

For this, if `CMD` spans multiple lines, eg:
```
CMD="touch $A \
  fileB"
```

I get:
```
[CMD]touch fileA \ 
  fileB
main.sh: line 11: fileB: command not found
```

How would I modify CMD accordingly? This is useful for me because bioinformatics commands are very long.

**Comment by fxquah** [comment]
Also, what does ``[[ "$DRY" -eq 1 ]]`` actually do?

**Comment by shouldsee** [comment]
- Multiline only in the script: here the command can be written in multilline during editing but actually become a single line command when documented. I think this is the most robust way.


```bash
#!/bin/bash

A=fileA
CMD="touch $A \
fileB; \
echo DONE
"
### notice fileB followed by ";"

### This is the snippet that I reuse.
### It always echoes CMD, but skip its executation if $DRY is set to 1
echo "[CMD]$CMD"
[[ "$DRY" -eq 1 ]] || eval "$CMD"
```

- Actualy multiline command does not fit with `eval "$CMD"` .

```bash
#!/bin/bash

A=fileA
CMD="touch $A 
fileB; 
echo DONE
"

echo "[CMD]$CMD"
[[ "$DRY" -eq 1 ]] || eval "$CMD" ### this will error
```

**Comment by shouldsee** [comment]
`[[ "$DRY" -eq 1 ]] || eval "$CMD"` test whether DRY==1, and evaluate things afterwards if the test fails. `||` chains command until the one that is successful, whereas `&&` chains command until the one that failed.

**Comment by fxquah** [comment]
I might have found a solution. I will test it on my pipelines tomorrow morning. In a file called `main.sh':
```
#!/bin/bash

# the space between $A and \ must be there
A=fileA
CMD="touch $A \
fileB"

#prints the command run onto the terminal
echo "the command ran was: ${CMD}"   # can also redirect into log file with >

#then runs the command itself
echo "${CMD}" | bash

#show what files are in the directory
ls
```

Output:
```
$bash main.sh
the command ran was: touch fileA fileB
fileA	fileB	main.sh
$
```

**Comment by shouldsee** [comment]
Hi Fu,

Please note there is a difference between

format A

```

CMD="touch $A \
fileB"

```

And

format B
```

CMD="touch $A 
fileB"

```

whereas format A creates a single line variable $CMD, format B creates a true multiline variable $CMD. In the single line scenario the old solution would still work and there is no need to change. In the multiline scenario your solution unfortunately did not work, that is, the following script would raise errors


```bash
#!/bin/bash

# the space between $A and \ must be there
A=fileA

##### Note the "\" has been omitted to create an actual multiline $CMD
CMD="touch $A
fileB"

#prints the command run onto the terminal
echo "the command ran was: ${CMD}"   # can also redirect into log file with >

#then runs the command itself
echo "${CMD}" | bash

#show what files are in the directory
ls
```

The pipeing through bash "| bash"  is an interestnig solution. This means you can modify the behaviour of the $CMD by appending -x or -e. 

```bash
echo "${CMD}" | bash -x
### or 
echo "${CMD}" | bash -e
```

But be AWARE this piping would also create variable scope problem, revealed by a following test
test.sh
```bash
#!/bin/bash


CMD="bash ./t1.sh"

VAR=111 ### declare a non-global var
echo "${CMD}" | bash -x

export VAR=111 ### declare a global var
echo "${CMD}" | bash -x
```
t1.sh
```bash
echo VAR=$VAR

```



Regards
Feng

**Comment by tavareshugo** [comment]
really cool solution of piping command to `bash` @fxquah . And the `-x` or `-e` options suggested by @shouldsee  make it a one-liner :smile: 

Not sure if this is useful, but if you wanted to _both_ output the command to stdout _and_ to a log file, you could use the rather exotic [`tee`](https://en.wikipedia.org/wiki/Tee_(command)):

```
echo "touch file1 file2" | tee -a out.log | bash -x
# or, in this case equivalente
echo "touch file1 file2" | bash -x | tee -a out.log
```

Of course that in this case you can always achieve the same with an extra `echo`.

**Comment by shouldsee** [comment]
@tavareshugo Nice change! Though `bash -x` will always have the subshell variable scope problem that caused me to avoid bash now...

I also wrote an ancient function along the same line that wraps the call to a particular command. I probably should put `eval` when actually evaluate the result. 

https://github.com/shouldsee/synoBio/blob/b4c6735e6b89a6d3c5f1bf728766ca7b9f5c3263/src/util.sh#L535

```bash
function datefloat() {
  date +%s.%N
}
export -f datefloat

runWithTimeLog()
{
    # REF: https://unix.stackexchange.com/questions/70653/increase-e-precision-with-usr-bin-time-shell-command
    local TIMEFORMAT='%3R'
    local CMD="$@"
    local ALI=${ALI:-testALI}
    local PROG=${PROG:-testPROG}
    local T0 T1 Tdiff
    local SELF=${SELF:-testScript.sh}
    
#     echo $CMD
    [[ $DRY -eq 1 ]] || {
        T0=`datefloat`
        $CMD 2>&1 | tee ${ALI}.${PROG}.log 
        T1=`datefloat`
        Tdiff=`echo $T1 - $T0 | bc`
        echo $SELF,$Tdiff,\"$CMD\"
    }
}
export -f runWithTimeLog

runWithTimeLog sleep 1
ls -lhtr
```
